third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs

```diff
@@ -0,0 +1,1453 @@
+// Memory Safety Analysis
+
+// -------------------------------------------------------------------------------------------------
+// Program Analysis Domain
+
+/// Lifetime graph nodes represent information about memory locations, as well as nodes derived
+/// from that locations.
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct LifetimeNode {
+    /// Memory location associated with this node.
+    location: MemoryLocation,
+    /// Outgoing edges to descendants.
+    descendants: SetDomain<BorrowEdge>,
+    /// Backlinks to ancestors.
+    ancestors: SetDomain<LifetimeLabel>,
+}
+
+/// A label for a lifetime node.
+#[derive(PartialOrd, Ord, PartialEq, Eq, Clone, Copy, Debug)]
+struct LifetimeLabel(usize);
+
+/// A memory location, either a global in storage or a local on the stack.
+#[derive(PartialOrd, Ord, PartialEq, Eq, Clone, Debug)]
+enum MemoryLocation {
+    /// The associated memory is in global storage.
+    Global(QualifiedInstId<StructId>),
+    /// The associated memory is a local on the stack.
+    Local(TempIndex),
+    /// The associated memory has been reused, but the old value is still bound here to this node.
+    /// This happens after an update to the location.
+    Replaced,
+}
+
+/// Represents an edge in the lifetime graph.
+#[derive(PartialOrd, Ord, PartialEq, Eq, Clone, Debug)]
+struct BorrowEdge {
+    /// The kind of borrow edge.
+    kind: BorrowEdgeKind,
+    /// Whether this edge is a mutable borrow
+    is_mut: bool,
+    /// A location associated with the borrow. For Skip edges (see below), no location may be
+    /// present.
+    loc: Option<Loc>,
+    /// Target of the edge.
+    target: LifetimeLabel,
+}
+
+/// The different type of edges.
+#[derive(PartialOrd, Ord, PartialEq, Eq, Clone, Debug)]
+enum BorrowEdgeKind {
+    /// Borrows the local at the MemoryLocation in the source node.
+    BorrowLocal,
+    /// Borrows the global at the MemoryLocation in the source node.
+    BorrowGlobal,
+    /// Borrows a field from a reference.
+    BorrowField(FieldId),
+    /// Calls an operation, where the incoming references are used to derive outgoing references.
+    Call(Operation),
+    /// The `Skip` edge is used for graph composition and glues two graph nodes together via
+    /// connecting them. For more details see the implementation of the `LifetimeDomain`
+    /// join operator.
+    Skip,
+}
+
+impl LifetimeLabel {
+    /// Creates a new, unique and stable, life time label based on a code offset and
+    /// a qualifier to distinguish multiple labels at the same code point.
+    /// Since the program analysis could run fixpoint loops, we need to ensure that
+    /// these labels are the same in each iteration.
+    fn new(code_offset: CodeOffset, qualifier: u8) -> LifetimeLabel {
+        LifetimeLabel(((code_offset as usize) << 8) | (qualifier as usize))
+    }
+
+    /// Creates a new, unique and stable, lifetime label from two other labels.
```

Comment:
I am assuming we do not care about executing on 32-bit architectures? Would it be better to be explicit and use `u64` instead of `usize` inside LifetimeLabel?